File: src/http/safe-external-url.ts
export function safePublicHttpsUrl(value: string | undefined): string | undefined {
  if (!value) {
    return undefined;
  }

  try {
    const url = new URL(value);
    if (url.protocol !== "https:" || isLocalOrPrivateHost(url.hostname)) {
      return undefined;
    }

    url.username = "";
    url.password = "";
    url.hash = "";
    return url.toString();
  } catch {
    return undefined;
  }
}

export function safeEbayItemUrl(value: string | undefined): string | undefined {
  const safeUrl = safePublicHttpsUrl(value);
  if (!safeUrl) {
    return undefined;
  }

  const url = new URL(safeUrl);
  const hostname = url.hostname.toLocaleLowerCase("en-GB");
  if (hostname !== "ebay.co.uk" && hostname !== "www.ebay.co.uk" && hostname !== "ebay.com" && hostname !== "www.ebay.com") {
    return undefined;
  }

  url.search = "";
  return url.toString();
}

export function safeEbayImageUrl(value: string | undefined): string | undefined {
  const safeUrl = safePublicHttpsUrl(value);
  if (!safeUrl) {
    return undefined;
  }

  const url = new URL(safeUrl);
  const hostname = normalizedHostname(url.hostname);
  if (!isTrustedEbayImageHost(hostname)) {
    return undefined;
  }

  return url.toString();
}

function isTrustedEbayImageHost(hostname: string): boolean {
  return hostname === "ebayimg.com" || hostname.endsWith(".ebayimg.com") || hostname === "ebaystatic.com" || hostname.endsWith(".ebaystatic.com");
}

function isLocalOrPrivateHost(hostname: string): boolean {
  const host = normalizedHostname(hostname).replace(/^\[(.*)\]$/, "$1");
  if (
    host === "localhost" ||
    host.endsWith(".localhost") ||
    host.includes(":")
  ) {
    return true;
  }

  const ipv4Host = host.startsWith("::ffff:") ? host.slice("::ffff:".length) : host;
  const parts = ipv4Host.split(".").map(Number);
  if (parts.length !== 4 || parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) {
    return false;
  }

  const [first, second] = parts;
  return (
    first === 10 ||
    first === 127 ||
    (first === 169 && second === 254) ||
    (first === 172 && second >= 16 && second <= 31) ||
    (first === 192 && second === 168)
  );
}

function normalizedHostname(hostname: string): string {
  return hostname.toLocaleLowerCase("en-GB");
}

File: macos/Goggler/Goggler/Networking/SafeExternalURL.swift
import Foundation

/// Direct Swift port of `src/http/safe-external-url.ts`'s `safeEbayImageUrl` —
/// HTTPS-only, restricted to eBay's actual image CDN hosts, rejects
/// local/private hosts. The web app validates client-side even though the
/// backend already returns eBay-sourced URLs; the same defense-in-depth
/// applies here, arguably more so — a native app's outbound image request
/// isn't sandboxed by browser-style mixed-content/CORS protections the way
/// a web page's `<img>` tag is, so an unvalidated URL could be used to
/// probe internal network addresses from the user's own Mac.
func safeEbayImageURL(_ value: String?) -> URL? {
    guard let value,
          var components = URLComponents(string: value),
          components.scheme == "https",
          let host = components.host
    else {
        return nil
    }

    let normalizedHost = host.lowercased()
    guard !isLocalOrPrivateHost(normalizedHost) else { return nil }
    guard isTrustedEbayImageHost(normalizedHost) else { return nil }

    // Matches the TS original's stripping of userinfo/hash — not a known
    // exploit path for AsyncImage (a plain GET, no display-text use of the
    // URL), but free to do and keeps the two implementations in lockstep.
    components.user = nil
    components.password = nil
    components.fragment = nil

    return components.url
}

private func isTrustedEbayImageHost(_ hostname: String) -> Bool {
    hostname == "ebayimg.com" || hostname.hasSuffix(".ebayimg.com")
        || hostname == "ebaystatic.com" || hostname.hasSuffix(".ebaystatic.com")
}

private func isLocalOrPrivateHost(_ hostname: String) -> Bool {
    if hostname == "localhost" || hostname.hasSuffix(".localhost") || hostname.contains(":") {
        return true
    }

    let parts = hostname.split(separator: ".", omittingEmptySubsequences: false).map { Int($0) }
    guard parts.count == 4, let first = parts[0], let second = parts[1],
          parts.allSatisfy({ if let value = $0 { return (0...255).contains(value) } else { return false } })
    else {
        return false
    }

    return first == 10
        || first == 127
        || (first == 169 && second == 254)
        || (first == 172 && (16...31).contains(second))
        || (first == 192 && second == 168)
}
